crates/consensus/src/transaction/envelope.rs

```diff
@@ -1038,6 +1038,257 @@ mod serde_from {
+                TxEnvelope::Eip4844(s) => Self::Eip4844(BorrowedSignedTxEip4844::from(s)),
+                TxEnvelope::Eip7702(s) => Self::Eip7702(BorrowedSignedTxEip7702::from(s)),
+            }
+        }
+    }
+    impl<'a> From<BorrowedTxEnvelope<'a>> for TxEnvelope {
+        fn from(env: BorrowedTxEnvelope<'a>) -> Self {
+            match env {
+                BorrowedTxEnvelope::Legacy(borrowed) => Self::Legacy(borrowed.into()),
+                BorrowedTxEnvelope::Eip2930(borrowed) => Self::Eip2930(borrowed.into()),
+                BorrowedTxEnvelope::Eip1559(borrowed) => Self::Eip1559(borrowed.into()),
+                BorrowedTxEnvelope::Eip4844(borrowed) => Self::Eip4844(borrowed.into()),
+                BorrowedTxEnvelope::Eip7702(borrowed) => Self::Eip7702(borrowed.into()),
+            }
+        }
+    }
+
+    /// Trait for converting to/from a bincode-friendly representation
+    pub trait SerdeBincodeCompat: Sized + 'static {
+        /// The borrowed bincode representation
+        type BincodeRepr<'a>: Serialize + Deserialize<'a>;
+
+        /// Returns the borrowed representation of `self`
+        fn as_repr(&self) -> Self::BincodeRepr<'_>;
+        /// Constructs `Self` from the borrowed representation
+        fn from_repr(repr: Self::BincodeRepr<'_>) -> Self;
+    }
+
+    impl SerdeBincodeCompat for TxEnvelope {
+        type BincodeRepr<'a> = BorrowedTxEnvelope<'a>;
+
+        fn as_repr(&self) -> Self::BincodeRepr<'_> {
+            BorrowedTxEnvelope::from(self)
+        }
+
+        fn from_repr(repr: Self::BincodeRepr<'_>) -> Self {
+            repr.into()
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::SerdeBincodeCompat;
+        use crate::transaction::TxEnvelope;
+        use arbitrary::Arbitrary;
+        use rand::Rng;
+        use serde::{Deserialize, Serialize};
+
+        /// Helper struct for testing bincode roundtrip
+        #[derive(Debug, Serialize, Deserialize)]
+        #[serde(bound(deserialize = "'de: 'a"))]
+        struct Data<'a> {
+            /// Bincode representation of a TxEnvelope.
+            transaction: <TxEnvelope as SerdeBincodeCompat>::BincodeRepr<'a>,
+        }
+
+        #[test]
+        fn test_tx_envelope_bincode_roundtrip() {
+            // Fill a buffer with random data
+            let mut bytes = [0u8; 1024];
+            rand::thread_rng().fill(&mut bytes);
+
+            // Generate a random TxEnvelope
+            let tx: TxEnvelope =
+                TxEnvelope::arbitrary(&mut arbitrary::Unstructured::new(&bytes)).unwrap();
+
+            // Convert to bincode representation
+            let data = Data { transaction: tx.as_repr() };
+
+            // Serialize and then deserialize
+            let encoded = bincode::serialize(&data).expect("Serialization should succeed");
+            let decoded: Data<'_> =
+                bincode::deserialize(&encoded).expect("Deserialization should succeed");
```

Comment:
@mattsse  Error on this line when running test `Deserialization should succeed: Custom("Bincode does not support Deserializer::deserialize_identifier")`

It seems to be related to how the `impl SerdeBincodeCompat for TxEnvelope` is now for `TxEnvelope`, which is an untagged enum, and it seems bincode requires tagged enums to know which variant to choose. So bincode calls `deserialize_identifier` and fails at that point. Could use some thoughts on how to refactor 🙏